sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollerFlux.java

```diff
@@ -0,0 +1,262 @@
+     * should again invoked. Once a subscription received a successful event from activation Mono then we want
+     * to cache it and replay it to any future subscriptions.
+     *
+     * The decorated Mono also handles concurrent calls to activation. Only one of them will be able to call
+     * activation and other subscriptions will keep resubscribing until it sees a cached response or get a chance
+     * to call activation as the one previously entered the critical section got an error on activation.
+     *
+     * @param activationOperation a supplier that provide a Mono upon subscription execute activation operation.
+     * @return
+     */
+    private Mono<PollResponse<T>> deferredActivationMono(final Supplier<Mono<T>> activationOperation) {
+        return Mono.defer(() -> {
+            if (this.activationResponse != null) {
+                return Mono.just(this.activationResponse);
+            }
+            if (this.guardActivationCall.compareAndSet(this, 0, 1)) {
+                final Mono<T> activationMono;
+                try {
+                    activationMono = activationOperation.get();
+                } catch (Throwable throwable) {
+                    return Mono.error(throwable);
+                }
+                return activationMono.map((T result) -> {
+                    this.activationResponse = new PollResponse<>(LongRunningOperationStatus.NOT_STARTED, result);
+                    return this.activationResponse;
+                })
+                .switchIfEmpty(Mono.defer(() -> {
+                    this.activationResponse = new PollResponse<>(LongRunningOperationStatus.NOT_STARTED, null);
+                    return Mono.just(this.activationResponse);
+                }))
+                .doOnError(throwable -> guardActivationCall.compareAndSet(this, 1, 0));
+            } else {
+                return Mono.empty();
+            }
+        }).repeatWhenEmpty((Flux<Long> longFlux) -> longFlux.concatMap(ignored -> Flux.just(true)));
+    }
+
+    /**
+     * Do the polling until it reaches a terminal state.
+     *
+     * @param activationResponse the response from activation operation
+     * @return a Flux that emits polling event.
+     */
+    private Flux<AsyncPollResponse<T, U>> pollingLoop(PollResponse<T> activationResponse) {
+        return Flux.using(
+            // Create a state per subscription
+            () -> new State<>(activationResponse),
+            // Do polling
+            // set|read in state as needed, reactor guarantee thread-safety of state object.
+            state -> Mono.defer(() -> pollOperation.apply(activationResponse, state.getLastResponse()))
+                .delaySubscription(getDelay(state.getLastResponse()))
+                .repeat()
+                .takeUntil(currentPollResponse -> currentPollResponse.getStatus().isComplete())
+                .onErrorResume(throwable -> {
```

Comment:
If the error is not recoverable, we continue to poll until the user cancels?

Reply:
the current assumption is - in such cases we expect the SDK implementer to return a `PollResponse` with failed status and its value as details of failure along with any service returned data. 

Looking at the current Poller's code comment - it says: "We expect Mono from pollOperation should never return Mono.error(). If any unexpected scenario happens in pollOperation, it should catch it and return a valid PollResponse."

I'm not sure this choice was based on some architecture review - @JonathanGiles @hemanttanwar do you remember?

Reply:
The current implementation of poll operation in KV Poller, 
handles the error and returns FAILED response to poller.